### DISCUSSION TITLE
Dynamic pausable/resumable download ?

### DISCUSSION THREAD

**Discussion by KaKi87** [fastify/fastify#2673] [Q&A]
Hello,

I have the file download endpoint :

```js
const { statSync, createReadStream } = require('fs');
fastify.get('/download/:fileId', (request, reply) => {
    const { fileId } = request.params;
    const { fileName, filePath } = getFile(fileId);
    if(!fileName || !filePath) return reply.code(404).send();
    reply
        .header('content-disposition', `inline; filename="${fileName}"`)
        .header('content-length', statSync(filePath).size)
        .send(createReadStream(filePath));
});
```

The download file works, but it's not pausable and not resumable.

How can I make it pausable and resumable ?

Thanks.

EDIT : note that I can't use `fastify-static` for this matter, the endpoint must be dynamic.

**Comment by mcollina** [comment]
You need to implement the spec for byte ranges yourself, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests.

Note that if you are serving a file from the filesystem fastify-static offers `reply.sendFile()` which would do it all for you.

**Reply by KaKi87** [reply]
Thanks for answering, but `reply.sendFile(filePath)` returns an HTTP 404 error

**Reply by mcollina** [reply]
how have you got fastify-static configured? what is your layout on disk?

**Reply by KaKi87** [reply]
I haven't, because I don't want to expose a directory full of files.

**Reply by L2jLiga** [reply]
`fastify-static` uses [`send`](https://github.com/pillarjs/send) under the hood, you can take a look on library to understand how you can use it to send files

**Reply by mcollina** [reply]
@KaKi87 you need to configure fastify-static with https://github.com/fastify/fastify-static#disable-serving.

**Reply by KaKi87** [reply]
Okay, sorry I forgot to say that, I do have fastify-static already configured, to expose a `/public` directory, for a web UI 😅

**Reply by KaKi87** [reply]
Okay, now I understand !

If I wouldn't be using `fastify-static` at all, I would do :

```js
fastify.register(require('fastify-static'), {
	root: `${__dirname}/downloads`,
	serve: false
});

fastify.get('/download/:fileId', (request, reply) => {
	// TODO declare fileName
	reply.sendFile(fileName);
});
```

But, because I'm already using `fastify-static`, I must do :

```js
fastify.register(require('fastify-static'), {
	root: `${__dirname}/public`
});

fastify.get('/download/:fileId', (request, reply) => {
	// TODO declare fileName & directoryPath
	reply.sendFile(fileName, directoryPath);
});
```

Got it, thanks !! :)

**Reply by Eomm** [reply]
Note that you could do it using context encapsulation.
There is always the way to register multiple times a plugin

```js
fastify.register(function(instance, opts, next){
  instance.register(require('fastify-static'), {
	root: `${__dirname}/public`
  });
})

fastify.register(function(instance, opts, next){
  instance.register(require('fastify-static'), {
	root: `${__dirname}/downloads`,
	serve: false
  });

  instance.get('/download/:fileId', (request, reply) => {
	reply.sendFile(fileName);
  });
})
```

**Reply by KaKi87** [reply]
Interesting!
It seems clearer so I'm very tempted to do it just for readability, but what about performance?
Thanks.

**Reply by Eomm** [reply]
Adding this overhead will cost a bit more of ram (cause the one extra context).
It should not impact performance at all for these kinds of tasks

**Reply by KaKi87** [reply]
How can I make `sendFile` use the original file name instead of the route's name ? Thanks.

**Reply by mcollina** [reply]
A quick query on google for "change file name http response" returned https://stackoverflow.com/questions/1741353/how-to-set-response-filename-without-forcing-save-as-dialog as the first answer. It's correct.

**Reply by KaKi87** [reply]
That's actually what I originally used (cf. original post) but I thought that `sendFile` would overwrite it.
Thanks
